onnxruntime/core/mlas/lib/dgemm.cpp

```diff
@@ -885,121 +845,40 @@ Return Value:
     // works okay for operations involving skinny matrices.
     //
 
+    ptrdiff_t ThreadsPerGemm = (TargetThreadCount + BatchSize - 1) / BatchSize;
```

Comment:
Still use the case that batch size is 4 and TargetThreadCount is 6. With this change, ThreadsPerGemm is 2 and TotalThreads/tasks is 8. For the first 6 task, cpu can be fully used but for the last 2 task, only 1/3 cpu is used. Could you make some perf test on this case and see whether it can get better or on-par performance than the baseline? If it doesn't, maybe we can fall back to the old strategy for the case that batch size is small.

Reply:
Test T5 saw no regression, most expensive matmul show 20% improvement. The jist of this change is to remove the barrier from one batch to another, which can be very expensive as inter thread communication easily reach millisecond level.

 I can think of a formula that would partition the job evenly for all the cases. Open to suggestions.

And with big.Little architecture, we need to retire this kind of "even" partition anyway, as the cores do not have equal computing power, and the barrier operation would become even more expensive.